src/model/trader.rs

```diff
@@ -212,33 +224,79 @@ pub fn settle_trade_on_counterparty(mut counterparty: Trader, offer: &Offer) ->
+        );
+    }
     // Settle inventories.
     counterparty.acquire(&Resource::Food, -1 * offer.food_delta());
     counterparty.acquire(&Resource::Water, -1 * offer.water_delta());
-    println!(
-        "Final inventory: Food: {}, Water: {}",
-        counterparty.forager.count(&Resource::Food),
-        counterparty.forager.count(&Resource::Water)
-    );
-
+    if core_config().simulation.VERBOSITY > 1 {
+        println!(
+            "Final inventory: Food: {}, Water: {}",
+            counterparty.forager.count(&Resource::Food),
+            counterparty.forager.count(&Resource::Water)
+        );
+    }
     counterparty
 }
 
 impl Agent for Trader {
     fn step(&mut self, state: &mut dyn krabmaga::engine::state::State) {
-        // all trades were facilitated in `before_step` function on the `board`
-        // next the agents choose an action given the agent_state post trading
+        let board = state.as_any_mut().downcast_mut::<Board>().unwrap();
+        if (board.step > 0) & board.has_trading {
+            // Traders state is fixed so can be used to retrieve trader
+            let mut traders = board.get_agents();
+            traders.shuffle(&mut board.rng);
+
+            // Execute trade if available.
+            if !self.offer().is_trivial() {
+                if !board.traded.contains_key(&self.id()) {
+                    // println!("Not traded yet");
+                    let offer = self.offer();
+                    for counterparty in &traders {
+                        // println!("Counterparty: {}", counterparty);
+                        let counterparty_id = counterparty.id();
+                        // If already traded, continue
+                        if board.traded.contains_key(&counterparty_id) {
+                            continue;
+                        }
+                        if counterparty_id != self.id()
+                            && counterparty.offer().matched(&offer)
+                            && (step_distance(&self.forager.pos, &counterparty.forager.pos)
+                                < core_config().trade.MAX_TRADE_DISTANCE)
+                        {
+                            if core_config().simulation.VERBOSITY > 1 {
+                                println!("Trade between: {} and {}", self, counterparty);
+                            }
+                            board.traded.insert(self.id(), Some(counterparty_id));
+                            board.traded.insert(counterparty.id(), Some(self.id()));
+                            *self = settle_trade_on_counterparty(*self, &offer.invert());
+                            // Break as trade has occurred
+                            break;
+                        }
+                    }
+                    // If no trade possible and not traded, set to None
+                    board.traded.entry(self.id()).or_insert(None);
+                } else if let Some(&Some(_)) = board.traded.get(&self.id()) {
```

Comment:
The design of the krABMaga [scheduler](https://docs.rs/krabmaga/latest/src/krabmaga/engine/schedule.rs.html#367-398) means I don't think we can easily get a mutable reference to the counterparty. Our current implementation uses `board.get_agents()` to give us "copies"/"snapshots" of the agents at the start of the state's timestep and not the ones the scheduler will call `.step()` on [here](https://docs.rs/krabmaga/latest/src/krabmaga/engine/schedule.rs.html#388). While this is a bit limiting (it would be nice to implement as you described) the upside is it's simpler if agent updates can only be done by themselves.

Two modifications given your comment:
- Good call on using a trader's trait method. I've added one that will `apply_offer()` and I've removed `settle_trade_on_counterparty` as it is not used elsewhere
- Calling `board.get_agents()` in each agent update is expensive as it loops over the whole board. So I've added a `current_traders` to the board so we only get them once from the `agent_grid` during `before_step()`

Reply:
Yep all clear, I think its a good way to do it.

Reply:
One final thing - why do we still need the `settle_trade` method? Is it not a bit dangerous to have it in the api given that it is unlikely a developer will be able to get a true mutable reference to two traders at the same time?